library/src/main/java/com/getstream/sdk/chat/view/MessageInputView.kt

```diff
@@ -0,0 +1,501 @@
+		} else {
+			messageInputController.onClickOpenSelectView(viewModel.channel, attachments, true)
+		}
+	}
+
+	// endregion
+	// region permission check
+	fun captureMedia(requestCode: Int, resultCode: Int, data: Intent?) {
+		if (requestCode == Constant.CAPTURE_IMAGE_REQUEST_CODE
+				&& resultCode == Activity.RESULT_OK) {
+			val imageFile = CaptureController.getCaptureFile(true)
+			val vieoFile = CaptureController.getCaptureFile(false)
+			if (imageFile == null && vieoFile == null) {
+				Utils.showMessage(context, context.getString(R.string.stream_take_photo_failed))
+				return
+			}
+			if (imageFile != null && imageFile.length() > 0) {
+				messageInputController.progressCapturedMedia(viewModel.channel, imageFile, true)
+				updateGallery(imageFile)
+			} else if (vieoFile != null && vieoFile.length() > 0) {
+				messageInputController.progressCapturedMedia(viewModel.channel, vieoFile, false)
+				updateGallery(vieoFile)
+			} else Utils.showMessage(context, context.getString(R.string.stream_take_photo_failed))
+		}
+	}
+
+	private fun updateGallery(outputFile: File) {
+		val scanIntent = Intent(Intent.ACTION_MEDIA_SCANNER_SCAN_FILE)
+		val contentUri = Uri.fromFile(outputFile)
+		scanIntent.data = contentUri
+		context.sendBroadcast(scanIntent)
+	}
+
+	/*Used for handling requestPermissionsResult*/
+	fun permissionResult(requestCode: Int,
+	                     permissions: Array<String>,
+	                     grantResults: IntArray) {
+		if (requestCode == Constant.PERMISSIONS_REQUEST) {
+			var storageGranted = true
+			var cameraGranted = true
+			var permission: String
+			var grantResult: Int
+			for (i in permissions.indices) {
+				permission = permissions[i]
+				grantResult = grantResults[i]
+				if (permission == Manifest.permission.CAMERA) {
+					cameraGranted = grantResult == PackageManager.PERMISSION_GRANTED
+				} else if (grantResult != PackageManager.PERMISSION_GRANTED) {
+					storageGranted = false
+				}
+			}
+			if (storageGranted && cameraGranted) {
+				messageInputController.onClickOpenBackGroundView(MessageInputType.ADD_FILE)
+				style.setCheckPermissions(true)
+			} else {
+				val message: String
+				message = if (! storageGranted && ! cameraGranted) {
+					context.getString(R.string.stream_both_permissions_message)
+				} else if (! cameraGranted) {
+					style.setCheckPermissions(true)
+					context.getString(R.string.stream_camera_permission_message)
+				} else {
+					style.setCheckPermissions(true)
+					context.getString(R.string.stream_storage_permission_message)
+				}
+				PermissionChecker.showPermissionSettingDialog(context, message)
+			}
+			messageInputController.configPermissions()
+		}
+	}
+
+	// endregion
+	// region listeners
+	protected fun setMessageSendListener(manager: MessageSendListener?) {
+		messageSendListener = manager
+	}
+
+	fun setPermissionRequestListener(l: PermissionRequestListener?) {
+		permissionRequestListener = l
+	}
+
+	private fun generateMessageID(): String {
```

Comment:
Message is generated at llc before it's sent.

Reply:
Good to know, BTW, it was implemented on that way on the previous code, and it has not been refactored yet. The idea is that `MessageInputView` will have a listener to notify a new message needs to be sent, and the ViewModel (Or the implementation that the developer provides) will handle it and use the LLC to send the message, no message-id will be generated internally on the view

Reply:
Cool.